### PR TITLE
Ice Thermal Storage Plant Comp Refactor

### PR DESCRIPTION
**Pull request by mitchute**
Pull request overview
---------------------

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by mitchute** [comment]
@Myoldmopar can you take a look at this sometime? The text diffs are expected, but the small math diffs are not as I noted offline. Also, the PTHPFixture failure doesn't fail locally, so I'm not sure what's going on there either.

**Comment by Myoldmopar** [comment]
The PTHP failure is a known Mac thing.  There is something that isn't being reinitialized in a clear_state somewhere, and it just _randomly_ pops up.  Nothing for you to worry about.  I'll take a look at the math diff.

File: src/EnergyPlus/IceThermalStorage.cc [hunk 1]
```diff
@@ -97,361 +96,207 @@ namespace IceThermalStorage {
...
-    int const DetIceInsideMelt(1);  // Inside melt system--charge starting with bare coil
-    int const DetIceOutsideMelt(2); // Outside melt system--charge from existing ice layer on coil
-
     // ITS parameter
-    Real64 const FreezTemp(0.0);       // Water freezing Temperature, 0[C]
-    Real64 const FreezTempIP(32.0);    // Water freezing Temperature, 32[F]
-    Real64 const TimeInterval(3600.0); // Time Interval (1 hr) [s]
-    int const ITSType_IceOnCoilInternal(1);
-    int const ITSType_IceOnCoilExternal(2);
+    Real64 const modFreezTemp(0.0);       // Water freezing Temperature, 0[C]
```
Comment (Myoldmopar) [inline src/EnergyPlus/IceThermalStorage.cc]: I mentioned this in the temperature source PR separately, so I know it's late here, but in general, the const module level variables are OK to leave as they are, without renaming.  They are fixed values, not related to state data from any single instance of an object, so they shouldn't cause any diff or refactoring issues.

File: src/EnergyPlus/IceThermalStorage.cc [hunk 2]
```diff
@@ -97,361 +96,207 @@ namespace IceThermalStorage {
...
-        ITSOutletSetPointTemp = 0.0;
-        ITSCoolingRate = 0.0;
-        ITSCoolingEnergy = 0.0;
-        ChillerOutletTemp = 0.0;
-        CheckEquipName.clear();
-        IceStorage.deallocate();
-        IceStorageReport.deallocate();
-        DetIceStor.deallocate();
-        IceStorageTypeMap.deallocate();
+        NumSimpleIceStorage = 0;
```
Comment (Myoldmopar) [inline src/EnergyPlus/IceThermalStorage.cc:134]: The get-input flag probably needs to be reset in the clear_state here.

File: src/EnergyPlus/IceThermalStorage.cc [hunk 3]
```diff
@@ -97,361 +96,207 @@ namespace IceThermalStorage {
...
+            }
         }
+        // If we didn't find it, fatal
+        ShowFatalError("LocalSimpleIceStorageFactory: Error getting inputs for simple ice storage named: " + objectName); // LCOV_EXCL_LINE
+        // Shut up the compiler
+        return nullptr; // LCOV_EXCL_LINE
+    }
 
-        if (firstTime) {
+    PlantComponent *DetailedIceStorageData::factory(std::string const &objectName)
```
Comment (Myoldmopar) [inline src/EnergyPlus/IceThermalStorage.cc:161]: Interesting.  I didn't actually realize there were two totally separate objects in here.  No issues for now, though I wonder if the overall module could be cleaned up later to just use one object with small variations in terms of physics calculations.  This all looks fine though.

File: src/EnergyPlus/IceThermalStorage.cc [hunk 4]
```diff
@@ -850,17 +666,17 @@ namespace IceThermalStorage {
...
-                        DetIceStor(IceNum).OutletTemp = TempSetPt;
-                        DetIceStor(IceNum).TankOutletTemp = ToutNew;
-                        DetIceStor(IceNum).CompLoad = DetIceStor(IceNum).MassFlowRate * Cp * std::abs(TempIn - TempSetPt);
-                        DetIceStor(IceNum).TankMassFlowRate = DetIceStor(IceNum).CompLoad / (Cp * std::abs(TempIn - ToutNew));
-                        DetIceStor(IceNum).BypassMassFlowRate = DetIceStor(IceNum).MassFlowRate - DetIceStor(IceNum).TankMassFlowRate;
+                        this->OutletTemp = TempSetPt;
+                        this->TankOutletTemp = ToutNew;
+                        this->CompLoad = this->MassFlowRate * Cp * std::abs(TempIn - TempSetPt);
+                        this->TankMassFlowRate = this->CompLoad / (Cp * std::abs(TempIn - ToutNew));
+                        this->BypassMassFlowRate = this->MassFlowRate - this->TankMassFlowRate;
```
Comment (Myoldmopar) [inline src/EnergyPlus/IceThermalStorage.cc:680]: Everything looks OK in this function.

File: src/EnergyPlus/IceThermalStorage.cc [hunk 5]
```diff
@@ -886,517 +700,442 @@ namespace IceThermalStorage {
...
-            IceStorageReport(IceNum).IceFracRemain = 1.0;
-            IceStorageReport(IceNum).ITSCoolingRate = 0.0;
-            IceStorageReport(IceNum).ITSCoolingEnergy = 0.0;
-            IceStorageReport(IceNum).ITSChargingRate = 0.0;
-            IceStorageReport(IceNum).ITSChargingEnergy = 0.0;
-            IceStorageReport(IceNum).ITSmdot = 0.0;
-            IceStorageReport(IceNum).ITSInletTemp = 0.0;
-            IceStorageReport(IceNum).ITSOutletTemp = 0.0;
+            SimpleIceStorage(iceNum).MyLoad = 0.0;
+            SimpleIceStorage(iceNum).Urate = 0.0;
```
Comment (Myoldmopar) [inline src/EnergyPlus/IceThermalStorage.cc:790]: I see the `U` member variable is no longer initialized to zero here.  I'm sure there is a reason, just noting it.

File: src/EnergyPlus/IceThermalStorage.cc [hunk 6]
```diff
@@ -2408,172 +1710,75 @@ namespace IceThermalStorage {
...
-
     Real64 TempSItoIP(Real64 const Temp)
     {
-
-        // Return value
-        Real64 TempSItoIP;
-
-        TempSItoIP = (Temp * 9.0 / 5.0) + 32.0;
-        return TempSItoIP;
+        return (Temp * 9.0 / 5.0) + 32.0;
```
Comment (Myoldmopar) [inline src/EnergyPlus/IceThermalStorage.cc:1732]: Hmm.  I'm definitely curious why we need to convert to IP internally.  But that's nothing to do with this refactor, so no worries.

File: src/EnergyPlus/IceThermalStorage.hh [hunk 7]
```diff
@@ -54,120 +54,60 @@
...
+    };
+
+    struct ITSType
+    {
+        enum
+        {
+            IceOnCoilInternal,
+            IceOnCoilExternal
+        };
+    };
```
Comment (Myoldmopar) [inline src/EnergyPlus/IceThermalStorage.hh:96]: The enum inside a struct is a bit odd.  Are you able to just use a standard `enum class ITSType` for these?

**Review by Myoldmopar — COMMENTED** [review]
This looks really good.  I made a few comments and requested some very minor changes.

**Comment by Myoldmopar** [comment]
I ran the diffs locally.  The only diff I see is in hourly facility electricity, and it's super spotty, like maybe 5-10 hours of the run period, and the absolute diff is around e-9, so no big deal.  There may have been something subtle you changed to clean up an initialization, but I don't believe it necessitates investigation.

File: src/EnergyPlus/IceThermalStorage.cc (hunk 1 above)
Comment (mitchute) [reply src/EnergyPlus/IceThermalStorage.cc]: :+1:

File: src/EnergyPlus/IceThermalStorage.cc (hunk 2 above)
Comment (mitchute) [reply src/EnergyPlus/IceThermalStorage.cc:134]: OK; I'll take another pass and make sure these all in the clear_state.

File: src/EnergyPlus/IceThermalStorage.cc (hunk 3 above)
Comment (mitchute) [reply src/EnergyPlus/IceThermalStorage.cc:161]: Agreed. I'm not sure when the right time is to do these types of cleanups, but I haven't been worrying about it too much on the plant-comp refactor work.

File: src/EnergyPlus/IceThermalStorage.cc (hunk 6 above)
Comment (mitchute) [reply src/EnergyPlus/IceThermalStorage.cc:1732]: Yeah, there's some weird SI->IP->SI shenanigans going on in here.

File: src/EnergyPlus/IceThermalStorage.hh (hunk 7 above)
Comment (mitchute) [reply src/EnergyPlus/IceThermalStorage.hh:96]: I'll fix it up.

**Comment by Myoldmopar** [comment]
I pulled develop into this branch locally and built and compared results with develop for the 9 ice storage example files.  I got no table diffs, expected text diffs, and 4 files with small math diffs.  These diffs are sparse values of e-9 Joules in hourly output, so I don't think there's anything to spend time investigating here.  This is good to go in.  Thanks @mitchute.

**Pull request merged**